Cron/Invalidate.php

```diff
@@ -0,0 +1,132 @@
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @author Nosto Solutions Ltd <contact@nosto.com>
+ * @copyright 2019 Nosto Solutions Ltd
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD 3-Clause
+ *
+ */
+
+namespace Nosto\Tagging\Cron;
+
+use Exception;
+use Magento\Framework\Data\Collection\AbstractDb;
+use Nosto\Tagging\Logger\Logger;
+use Nosto\Tagging\Model\ResourceModel\Magento\Product\Collection;
+use Nosto\Tagging\Model\Service\Index as NostoServiceIndex;
+use Nosto\Tagging\Helper\Account as NostoAccountHelper;
+use Nosto\Tagging\Model\ResourceModel\Product\Index\CollectionFactory as IndexCollectionFactory;
+use Magento\Store\Model\Store;
+
+/**
+ * Cronjob class that periodically invalidates Nosto indexed data for each of the store views
+ *
+ * @package Nosto\Tagging\Cron
+ */
+class Invalidate
+{
+    const MAX_UPDATED_AT_INTERVAL = 4;
+
+    /** @var Logger */
+    protected $logger;
+
+    /** @var IndexCollectionFactory */
+    private $indexCollectionFactory;
+
+    /** @var NostoServiceIndex */
+    private $nostoServiceIndex;
+
+    /** @var NostoAccountHelper */
+    private $nostoAccountHelper;
+
+    /**
+     * Invalidate constructor.
+     *
+     * @param Logger $logger
+     * @param IndexCollectionFactory $indexCollectionFactory
+     * @param NostoServiceIndex $nostoServiceIndex
+     * @param NostoAccountHelper $nostoAccountHelper
+     */
+    public function __construct(
+        Logger $logger,
+        IndexCollectionFactory $indexCollectionFactory,
+        NostoServiceIndex $nostoServiceIndex,
+        NostoAccountHelper $nostoAccountHelper
+    ) {
+        $this->logger = $logger;
+        $this->indexCollectionFactory = $indexCollectionFactory;
+        $this->nostoServiceIndex = $nostoServiceIndex;
+        $this->nostoAccountHelper = $nostoAccountHelper;
+    }
+
+    /**
+     * Executes cron job
+     * @throws Exception
+     */
+    public function execute()
+    {
+        $stores = $this->nostoAccountHelper->getStoresWithNosto();
+        foreach ($stores as $store) {
+            $productIndexCollection = $this->getCollection($store);
+            foreach ($productIndexCollection as $productIndex) {
```

Comment:
Introduce method `markCurrentItemsAsDirtyByStore ` in the collection directly (simlilar to https://github.com/Nosto/nosto-magento2/blob/feature/product-caching-index/Model/ResourceModel/Product/Index/Collection.php#L286) to avoid multiple updates to the table